egs/wsj/s5/steps/cleanup/append_eval_to_ctm.py

```diff
@@ -0,0 +1,396 @@
+#!/usr/bin/env python
+
+# Copyright 2016  Vimal Manohar
+# Apache 2.0
+
+from __future__ import print_function
+import sys, operator, argparse
+
+# Append to the CTM the Levenshtein alignment of 'hypothesis' and 'reference':
+# (i.e. the output of 'align-text' post-processed by 'wer_per_utt_details.pl')
+
+# The tags in the processed Levenshtein alignment are:
+#  'C' = correct
+#  'S' = substitution
+#  'D' = deletion
+#  'I' = insertion
+#  'EC' = silence and neighboring words are both correct
```

Comment:
On Tue, May 31, 2016 at 12:12 AM Daniel Povey <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/cleanup/append_eval_to_ctm.py
> <https://github.com/kaldi-asr/kaldi/pull/777#discussion_r65121717>:
>
> > +
> > +# Copyright 2016  Vimal Manohar
> > +# Apache 2.0
> > +
> > +from __future__ import print_function
> > +import sys, operator, argparse
> > +
> > +# Append to the CTM the Levenshtein alignment of 'hypothesis' and 'reference':
> > +# (i.e. the output of 'align-text' post-processed by 'wer_per_utt_details.pl')
> > +
> > +# The tags in the processed Levenshtein alignment are:
> > +#  'C' = correct
> > +#  'S' = substitution
> > +#  'D' = deletion
> > +#  'I' = insertion
> > +#  'EC' = silence and neighboring words are both correct
>
> > +# Copyright 2016 Vimal Manohar > +# Apache 2.0 > + > +from __future__
> import print_function > +import sys, operator, argparse > + > +# Append to
> the CTM the Levenshtein alignment of 'hypothesis' and 'reference': > +#
> (i.e. the output of 'align-text' post-processed by 'wer_per_utt_details.pl')
> > + > +# The tags in the processed Levenshtein alignment are: > +# 'C' =
> correct > +# 'S' = substitution > +# 'D' = deletion > +# 'I' = insertion >
> +# 'EC' = silence and neighboring words are both correct I don't understand
> the naming convention here, why E? I believe E was because silence was
> represented as <eps> in the CTM.
> It might be better to use 'S' as the prefix for silence-- this would be
> more obvious to readers. Your documentation says that the script "Append[s]
> to the CTM the Levenshtein alignment of 'hypothesis' and 'reference'". This
> would seem to be inconsistent with the notion that it's adding new CTM
> lines corresponding to deletions. Be precise! You should explain at this
> point what the precedence is among the types of
> error. For instance, del sil sub, what tag does that put on the silence?
> This is explained in the comments above the later functions. But I can add
> a line here.
> This comment is like an interface-level documentation (if this were a
> header file). Having documentation inside the implementation does not
> remove the obligation to have documentation of the interface. If some parts
> would take too long to explain you can refer to other parts of the code.
> What does it mean for a neighboring word to be a deletion, given that a
> deletion does not appear in a transcript? You might need to develop more
> precise language for this (or explain further below). Does you add extra
> lines to the CTM for deletions? Does 'silence' mean any phone that is in
> phones/silence.txt, or just the optional silence? Deletion is found from
> using align-text information. A silence token <eps> is marked as a deletion
> if there is a silence present or otherwise a silence token of length 0 is
> added to the CTM. In the final processed CTM with 'ED', 'EI' etc. an ED
> will always be next to a D.
> I think you need to distinguish between silence tokens that represent
> actual silence, and <eps> which is a placeholder used in alignments. I
> don't think you should be referring to the <eps> that arise from the
> alignment, as 'silence' in any context-- that's not at all what they mean.
> We do under some circumtances use <eps> as the word label for silences in
> the ctm where there is no corresponding label in the graph, but I don't
> think it's a good idea to confuse these with the <eps> tokens that arise
> from the alignment. Making this distinction may require some non-cosmetic
> changes to the script.
>
<eps> is what is obtained from  nbest-to-ctm --print-silence=true. I do not
use phone alignments anywhere for segmentation. Do you mean this to also be
used?

Is it really necessary to add this information at this point, given that
> … <#m_-5635296113375199773_>
> much of it is easy to work out from the CTM? It simplifies later
> segmentation script. Some things like handling deletions are more
> complicated. Be clear about what you mean by an OOV-- presumably a
> reference word that does not appear in words.txt. Try to be explicit what
> you mean about 'words' here (e.g. in 'neighboring words'), given that there
> are two word sequences involved. — You are receiving this because you
> commented. Reply to this email directly, view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/777/files/b523f9d63f1a28d24e4a59eff59c6abec472147a#r65121245>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/ADJVu4qfDFNdFOFSi24pvOHn1MpEW5PTks5qG7I7gaJpZM4IeUC0>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/777/files/b523f9d63f1a28d24e4a59eff59c6abec472147a#r65121717>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEATV0QUf6toUtA3MWOtEMWQxm353wItks5qG7UUgaJpZM4IeUC0>
> .
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University